src/Microsoft.DotNet.Build.VSTSBuildsApi/VSTSBuildClient.cs

```diff
@@ -0,0 +1,234 @@
+        /// check VSTS for a matching definition name, if present, update that definition with the
+        /// local build definition, otherwise create a new definition.
+        /// </summary>
+        /// <param name="definition">VSTS build definition JSON object</param>
+        /// <returns>Created or updated build definition id</returns>
+        private async Task<string> CreateOrUpdateDefinitionAsync(JObject definition)
+        {
+            var key = string.Join("_", _collectionIdentifier, definition["name"].ToString());
+            definition["name"] = key;
+            var vstsDefinitions = await VstsRetrieveDefinitionsListByNameAndPathAsync(definition);
+
+            if (vstsDefinitions.Count == 0)
+            {
+                /* Create */
+                var vstsDefinition = await VstsCreateDefinitionAsync(definition);
+                return vstsDefinition["id"].ToString();
+            }
+            else if(vstsDefinitions.Count == 1)
+            {
+                var vstsDefinition = await VstsRetrieveDefinitionByIdAsync(vstsDefinitions[0]);
+
+                /* Update */
+                if (!IsSameContents(definition, vstsDefinition))
+                {
+                    definition["revision"] = vstsDefinition["revision"];
+                    definition["id"] = vstsDefinition["id"];
+                    definition["uri"] = vstsDefinition["uri"];
+                    definition["path"] = vstsDefinition["path"];
+                    definition["createdDate"] = vstsDefinition["createdDate"];
+                    definition["url"] = vstsDefinition["url"];
+                    definition["comment"] = "Automated update";
+
+                    vstsDefinition = await VstsUpdateDefinitionAsync(definition);
+                }
+                return vstsDefinition["id"].ToString();
+            }
+            else
+            {
+                throw new Exception(string.Format("Obtained multiple {0} definitions with the same 'name' ({1}) and 'path' ({2}) properties.  This should not be possible.", vstsDefinitions.Count, vstsDefinitions[0]["name"].ToString(), vstsDefinitions[0]["path"].ToString()));
+            }
+        }
+
+        private JObject GetDefinition(Stream stream)
+        {
+            using (StreamReader streamReader = new StreamReader(stream))
+            using (JsonTextReader jsonReader = new JsonTextReader(streamReader))
+            {
+               return (JObject)JToken.ReadFrom(jsonReader);
+            }
+        }
+
+        /// <summary>
+        /// Update VSTS Build Definition
+        /// </summary>
+        /// <param name="definition"></param>
+        /// <returns>updated definitions id</returns>
+        private async Task<JObject> VstsUpdateDefinitionAsync(JObject definition)
+        {
+            string requestUri = $"DefaultCollection/DevDiv/_apis/build/definitions/{definition["id"].ToString()}?api-version=2.0";
```

Comment:
I preferred to derive this from the build definitions "url", because it is possible that we would want this to apply to other accounts, like "mseng".